File: Jim/src/sk/fiit/jim/agent/highskill/WalkOld.java
package sk.fiit.jim.agent.highskill;

import sk.fiit.jim.agent.AgentInfo;
import sk.fiit.jim.agent.models.AgentModel;
import sk.fiit.jim.agent.models.EnvironmentModel;
import sk.fiit.jim.agent.models.WorldModel;
import sk.fiit.jim.agent.moves.LowSkill;
import sk.fiit.jim.agent.moves.LowSkills;
import sk.fiit.jim.agent.skills.HighSkill;
import sk.fiit.robocup.library.geometry.Angles;
import sk.fiit.robocup.library.geometry.Vector3D;


/**
 * 
 * Reimplemented from Ruby to Java by Stefan Horvath
 *
 */
public class WalkOld extends HighSkill {
	
	AgentInfo agentInfo = AgentInfo.getInstance();
	AgentModel agentModel = AgentModel.getInstance();
	
	Angles rightRange = new Angles(1.5*Math.PI, (11.0/6.0)*Math.PI);
	Angles leftRange = new Angles((1.0/6.0)*Math.PI, 0.5*Math.PI);
	Angles leftStrafeRange =  new Angles((5.0/12.0)*Math.PI, (0.5*Math.PI));
	Angles rightStrafeRange =  new Angles(1.5*Math.PI, ((19.0/12.0)*Math.PI));
	double strafeDistance = 1.5;
	Vector3D target = null;
	
	public WalkOld(Vector3D target) {
		super();
		this.target = target;
	}

	@Override
	public LowSkill pickLowSkill() {
		Vector3D targetPosition = null;
		if (target.equals(targetPosition = WorldModel.getInstance().getBall().getPosition()))
			targetPosition = WorldModel.getInstance().getBall().getRelativePosition();
		else
			targetPosition = WorldModel.getInstance().getBall().getPosition();
		targetPosition = targetPosition.setZ(0.0);
		if (closeEnought(targetPosition))
			return null;
		if (EnvironmentModel.beamablePlayMode())
			return null;
		if (agentModel.isOnGround())
			return null;
		if (ballUnseen() > 3)
			return null;
		if (straight(targetPosition))
			return LowSkills.get("walk_forward");
		if (leftAndDistant(targetPosition))
			return LowSkills.get("turn_left_cont_20");
		if (rightAndDistant(targetPosition))
			return LowSkills.get("turn_right_cont_20");
		return null;
	}
	
	public boolean closeEnought(Vector3D targetPosition) {
		return targetPosition.getR() < 0.0;
	}
	
	public boolean straight(Vector3D targetPosition) {
		return Angles.angleDiff(targetPosition.getPhi(), 0.0) < Math.PI / 6.0;
	}
	
	public boolean back(Vector3D targetPosition) {
		return Angles.angleDiff(targetPosition.getPhi(), Math.PI) < Math.PI / 6.0;
	}
	
	public boolean backAndRight(Vector3D targetPosition) {
		return (Angles.angleDiff(targetPosition.getPhi(), Math.PI) < Math.PI / 6.0) && (rightRange.include(Angles.normalize(targetPosition.getPhi())));
	}
	
	public boolean backAndLeft(Vector3D targetPosition) {
		return (Angles.angleDiff(targetPosition.getPhi(), Math.PI) < Math.PI / 6.0) && (leftRange.include(Angles.normalize(targetPosition.getPhi())));
	}
	
	public boolean leftAndClose(Vector3D targetPosition) {
		agentInfo.loguj(targetPosition.toString());
		return ((targetPosition.getR() <= strafeDistance) && leftStrafeRange.include(Angles.normalize(targetPosition.getPhi())));
	}
	
	public boolean rightAndClose(Vector3D targetPosition) {
		return ((targetPosition.getR() <= strafeDistance) && rightStrafeRange.include(Angles.normalize(targetPosition.getPhi())));
	}
	
	public boolean leftAndDistant(Vector3D targetPosition) {
		return leftRange.include(Angles.normalize(targetPosition.getPhi()));
	}
	
	public boolean rightAndDistant(Vector3D targetPosition) {
		return rightRange.include(Angles.normalize(targetPosition.getPhi()));
	}
	
	public double ballUnseen() {
		return EnvironmentModel.SIMULATION_TIME - WorldModel.getInstance().getBall().getLastTimeSeen();
	}
	
	@Override
	public void checkProgress() throws Exception {
	}

}

File: Old_xml_and_scripts/Jim/high_skills/new-18.5.2013/walkOld.rb
require 'high_skills/ruby_high_skill.rb'
require 'plan/plan.rb'
include_class Java::sk.fiit.robocup.library.geometry.Angles
include_class Java::java.lang.Math
include_class Java::sk.fiit.jim.agent.models.AgentModel
include_class Java::sk.fiit.jim.agent.models.EnvironmentModel

class WalkOld < RubyHighSkill
  
  #270 - 330 degrees
  @@right_range = (1.5*Math::PI)..((11.0/6.0)*Math::PI)
  #30 - 90 degrees
  @@left_range = ((1.0/6.0)*Math::PI)..(0.5*Math::PI)
  #75 - 90 degrees
  @@left_strafe_range = ((5.0/12.0)*Math::PI..(0.5*Math::PI))
  #270 - 285 degrees
  @@right_strafe_range = (1.5*Math::PI..(19.0/12.0)*Math::PI)
  @@strafe_distance = 1.5
  
  
  def initialize target
    super()
    @target = target
  end
  
  def pickLowSkill
    target_position = @target
    if @target == :ball then
    	target_position = (ball relative position)
    else
    	target_position = AgentModel.getInstance.relativize @target
    end
    
    #puts 'moviing this>'
    #puts target_position
    
    target_position = target_position.set_z(0.0)
    #puts 'going to'
    #puts target_position
    #puts 'going to'
    return nil if close_enough?(target_position)
    return nil if EnvironmentModel.beamablePlayMode
    #puts AgentModel.getInstance.position.getZ
	#return get_skill("stand_back") if me.on_ground?
	return nil if me.on_ground?
	
	return nil if ball_unseen > 3
    return get_skill("walk_forward") if straight?(target_position)
    #return get_skill("turn_left_90") if back_and_left?(target_position)
    #return get_skill("turn_right_90") if back_and_right?(target_position)
    #return get_skill("step_left") if left_and_close?(target_position)
    #puts '4'
    #return get_skill("step_right") if right_and_close?(target_position)
    #puts '5'
    return get_skill("turn_left_cont_20") if left_and_distant?(target_position)
    return get_skill("turn_right_cont_20") if right_and_distant?(target_position)
  end
  
  def close_enough? target_position
    target_position.r < 0.0
  end
  
  def straight? target_position
    Angles.angle_diff(target_position.phi, 0.0) < Math::PI / 6.0 
  end
  
  #def on_ground?
  #  AgentModel.getInstance.position.getZ < 0.5 
  #end
  
  def back? target_position
    Angles.angle_diff(target_position.phi, Math::PI) < Math::PI / 6.0
  end
  
  def back_and_right? target_position
    Angles.angle_diff(target_position.phi, Math::PI) < Math::PI / 6.0 and @@right_range.include?(Angles.normalize(target_position.phi))
  end
  
  def back_and_left? target_position
    Angles.angle_diff(target_position.phi, Math::PI) < Math::PI / 6.0 and @@right_range.include?(Angles.normalize(target_position.phi))
  end
  
  def left_and_close? target_position
    puts target_position
    (target_position.r <= @@strafe_distance) and @@left_strafe_range.include?(Angles.normalize(target_position.phi))
  end
  
  def right_and_close? target_position
    (target_position.r <= @@strafe_distance) and @@right_strafe_range.include?(Angles.normalize(target_position.phi))
  end
  
  def left_and_distant? target_position
    @@left_range.include?(Angles.normalize(target_position.phi))
  end
  
  def right_and_distant? target_position
    @@right_range.include?(Angles.normalize(target_position.phi))
  end
  
  def checkProgress
#    raise "Fallen!" if me.on_ground?
  end
  
  def abs number
    Math.abs number
  end
  
  def ball_unseen
    EnvironmentModel.SIMULATION_TIME - WorldModel.getInstance.getBall.getLastTimeSeen()
  end
end

#in order not to defer the startup cost
WalkOld.new :ball